pkg/crc/machine/sync.go

```diff
@@ -0,0 +1,119 @@
+package machine
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/code-ready/crc/pkg/crc/logging"
+	"github.com/code-ready/machine/libmachine/state"
+	"golang.org/x/sync/semaphore"
+)
+
+const startCancelTimeout = 15 * time.Second
+
+type Synchronized struct {
+	startLock      *semaphore.Weighted
+	stopDeleteLock *semaphore.Weighted
+	startCancel    context.CancelFunc
+	underlying     Client
+}
+
+func NewSynchronizedMachine(machine Client) *Synchronized {
+	return &Synchronized{
+		startLock:      semaphore.NewWeighted(1),
+		stopDeleteLock: semaphore.NewWeighted(1),
+		underlying:     machine,
+	}
+}
+
+func (s *Synchronized) IsProgressing() bool {
+	if !s.startLock.TryAcquire(1) {
+		return true
+	}
+	defer s.startLock.Release(1)
+	if !s.stopDeleteLock.TryAcquire(1) {
+		return true
+	}
+	defer s.stopDeleteLock.Release(1)
+	return false
+}
+
+func (s *Synchronized) Delete() error {
+	cleanup, err := s.lockForStopOrDelete()
+	defer cleanup()
+	if err != nil {
+		return err
+	}
+	return s.underlying.Delete()
+}
+
+func (s *Synchronized) Start(ctx context.Context, startConfig StartConfig) (*StartResult, error) {
+	if !s.startLock.TryAcquire(1) {
+		return nil, errors.New("cluster is starting")
+	}
+	defer s.startLock.Release(1)
+	ctx, s.startCancel = context.WithCancel(ctx)
```

Comment:
`s.startCancel` is never reset to nil, not sure if this matters or not.

Reply:
Yes added a missing `defer`. Also, cancelFunc can be called multiple times, so no issues.